mathesar/utils/forms.py

```diff
@@ -0,0 +1,113 @@
+from uuid import uuid4
+from collections import defaultdict
+
+from django.db import transaction
+
+from db.forms import get_oid_col_info_map
+from db.roles import get_current_role_from_db
+from mathesar.models.base import (
+    Form, FormField, ConfiguredRole, UserDatabaseRoleMap, ColumnMetaData
+)
+from mathesar.rpc.columns.metadata import ColumnMetaDataBlob
+
+
+def get_submit_role(user, database_id, submit_role_id=None):
+    user_dbrm = UserDatabaseRoleMap.objects.get(user=user, database__id=database_id)
+    with user_dbrm.connection as conn:
+        submit_role = (
+            ConfiguredRole.objects.get(id=submit_role_id) if submit_role_id else None
+            or user_dbrm.configured_role
+        )
+        current_role = get_current_role_from_db(conn)
+        parent_names = {pr["name"] for pr in current_role["parent_roles"]}
+        assert (
+            user_dbrm.configured_role == submit_role
+            or user_dbrm.user.is_superuser
+            or submit_role.name in parent_names
+        ), "Insufficient privileges for selected submit_role"
+    return submit_role
+
+
+def get_oid_attnums_map(form_model):
+    oam = defaultdict(list)
+    fields = {field.key: field for field in form_model.fields.all()}
+    for field in fields.values():
+        table_oid = field.parent_field.target_table_oid if field.parent_field else form_model.base_table_oid
+        oam[table_oid].append(field.attnum)
+    return oam
+
+
+def get_field_col_info_map(user, form_model):
```

Comment:
Just realized a problem while reviewing #4554 . 

This shouldn't take the user as an arg. It should instead use the `submit_role` for that form. In particular, things like the permissions associated with columns need to reflect the form's role, not the current user don't they? In that case, just make this a method of the `Form` model. I.e., the call should be something like: "Dear `Form`, please tell me about yourself."

Reply:
Maybe _maybe_ you could have a method for non-public forms with no associated `submit_role` in the future that would involve the user's role, but I don't think that is needed at the moment.

Reply:
We _are_ switching user's configured role with `submit_role` before opening a connection.

Reply:
Then why is the user relevant at all?

Reply:
ok, added a connection property within the `Form` model itself.